public/v1/spec.yaml

```diff
@@ -14,30 +30,1942 @@ info:
+                        $ref: '#/components/schemas/StatsPacketsLoss'
+                  timings:
+                    type: array
+                    description: |
+                      Details for each sent packet. 
+                      All times are in milliseconds.
+                    items:
+                      type: object
+                      required:
+                        - rtt
+                      properties:
+                        rtt:
+                          $ref: '#/components/schemas/TimingPacketRtt'
+    FinishedHttpTestResult:
+      title: FinishedHttpTestResult
+      allOf:
+        - $ref: '#/components/schemas/BaseFinishedTestResult'
+        - type: object
+          required:
+            - rawHeaders
+            - rawBody
+            - headers
+            - resolvedAddress
+            - statusCode
+            - statusCodeName
+            - timings
+            - tls
+          properties:
+            rawHeaders:
+              type: string
+              description: The raw HTTP response headers.
+            rawBody:
+              type: string
+              description: The raw HTTP response body. Only the first 10 kb are returned.
+            headers:
+              type: object
+              description: The HTTP response headers.
+              additionalProperties:
+                type: string
+            statusCode:
+              $ref: '#/components/schemas/HttpStatusCode'
+            statusCodeName:
+              $ref: '#/components/schemas/HttpStatusCodeName'
+            resolvedAddress:
+              $ref: '#/components/schemas/ResolvedAddress'
+            timings:
+              type: object
+              required:
+                - total
+                - dns
+                - tcp
+                - tls
+                - firstByte
+                - download
+              properties:
+                total:
+                  $ref: '#/components/schemas/TimingHttpTotalNullable'
+                dns:
+                  $ref: '#/components/schemas/TimingHttpDnsNullable'
+                tcp:
+                  $ref: '#/components/schemas/TimingHttpTcpNullable'
+                tls:
+                  $ref: '#/components/schemas/TimingHttpTlsNullable'
+                firstByte:
+                  $ref: '#/components/schemas/TimingHttpFirstByteNullable'
+                download:
+                  $ref: '#/components/schemas/TimingHttpDownloadNullable'
+            tls:
+              oneOf:
+                - $ref: '#/components/schemas/TlsCertificate'
+                - type: 'null'
+    TestRawOutput:
+      type: string
+      description: |
+        The raw output can be presented to users but is not meant to be parsed clients.
+        Please use the individual values provided in other fields for automated processing.
+    BaseTestStatus:
+      type: string
+      description: The test status. Any value other than `in-progress` is final.
```

Comment:
I initially wrote it like that, but then we ended up with many nullable fields in `results` because `in-progress` and `failed` don't have them. I think it'll be easier to work with the types as they are now because once you check for `status: finished` it's clear which fields are there.

Not sure if you also checked the render version but this is how it looks now:
![image](https://github.com/jsdelivr/globalping/assets/6192491/25741fe6-e651-467d-8cfc-9d5b9b4db3a5)

Reply:
Good. I'm certainly open to reshaping the definitions once we have some real-use experience with them or user feedback. My current impression is this endpoint has so many possible states that it was quite difficult to describe it in a reasonably precise and not-too-verbose way.